examples/stm32h7/src/bin/signal_shared.rs

```diff
@@ -0,0 +1,125 @@
+use defmt_rtt as _;
+
+use panic_probe as _;
+
+use embassy::channel::Signal;
+use embassy::executor::Spawner;
+use embassy::time::{Duration, Timer};
+use embassy_stm32::Peripherals;
+
+// Expected output (with defmt-print):
+//
+// 0.000030 INFO  registering sender
+// 0.000091 INFO  main task doing other things
+// 0.000152 INFO  getting sender
+// 0.000213 INFO  waiting for signal
+// 1.000152 INFO  main task doing other things
+// 1.000213 INFO  signalling, counter: 0
+// 1.000335 INFO  signalled, counter: 0
+// 2.000213 INFO  main task doing other things
+// 2.000274 INFO  signalling, counter: 1         <-- receiver misses this
+// 3.000274 INFO  main task doing other things
+// 3.000457 INFO  signalling, counter: 2         <-- receiver misses this
+// 4.000335 INFO  main task doing other things
+// 4.000549 INFO  signalling, counter: 3         <-- receiver misses this
+// 5.000396 INFO  main task doing other things
+// 5.000640 INFO  signalling, counter: 4
+// 6.000427 INFO  waiting for signal
+// 6.000457 INFO  signalled, counter: 4
+// 6.000579 INFO  main task doing other things
+// 6.000762 INFO  signalling, counter: 5
+// 7.000640 INFO  main task doing other things
+// ...
+// 7.000854 INFO  signalling, counter: 6         <-- receiver misses this
+// 8.000701 INFO  main task doing other things
+// ...
+// 10.001129 INFO  signalling, counter: 9
+// 11.000579 INFO  waiting for signal
+// 11.000610 INFO  signalled, counter: 9
+// 11.000885 INFO  main task doing other things
+
+mod sender {
+    use super::info;
+    use super::Signal;
+    use super::{Duration, Timer};
+    use core::borrow::Borrow;
+
+    static SIGNAL: Signal<u32> = Signal::new();
+
+    #[embassy::task]
+    pub async fn my_sending_task() {
+        let mut counter: u32 = 0;
+
+        loop {
+            Timer::after(Duration::from_secs(1)).await;
+
+            info!("signalling, counter: {}", counter);
+            SIGNAL.signal(counter);
+
+            // receiving task gets a copy of the counter, at it's current value, before it is incremented below
+
+            counter = counter.wrapping_add(1);
+        }
+    }
+
+    pub fn init_sending_task() {
+        info!("registering sender");
+        super::receiver::register_sender(SIGNAL.borrow());
+    }
+}
+
+mod receiver {
+    use super::info;
+    use super::Signal;
+    use super::{Duration, Timer};
+    use core::cell::RefCell;
+    use embassy::blocking_mutex::raw::CriticalSectionRawMutex;
+    use embassy::blocking_mutex::Mutex;
+
+    // RefCell because Mutux::lock() give a immutable reference to the closure.
+    // Option because a default value needs to be provided for a static value.
+    // 'static lifetime for the reference because the signal needs to live forever.
+    // Mutex because otherwise you get a 'cannot be shared between threads safely' due to the static lifetime requirement.
+    static SIGNAL: Mutex<CriticalSectionRawMutex, RefCell<Option<&'static Signal<u32>>>> =
```

Comment:
This is very non-idiomatic. Signal is designed to not need extra mutexes, it already does mutexing internally.

If you want two mods to talk to each other without one poking into the other's statics, you can pass the signal as a task argument: `#[embassy::task] pub async fn my_receiving_task(signal: &'static Signal<u32>) { ..`

Reply:
The cleanest would probably be `main` having the `static` and passing `&'static Signal` to both tasks. Each task declares the resources it needs, and main "wires" them together.

Reply:
yeah, there's so many ways of doing it.  I was looking for a solution where main knows nothing about the internal needs of the other tasks other than they both might need an init() to be called on them.  Yes, I'm aware that I could pass the static from the `main` function, again the idea was to hide the internals.

Yes, agree that the need for an additional mutex is far from ideal.  The second mutex is only for initialisation and also so that task excution order doesn't matter.

Any other suggestions?

Reply:
> I was looking for a solution where main knows nothing about the internal needs of the other tasks

If main forgets to call `init_sending_task` or calls it in the wrong order, `my_receiving_task` will panic at the unwrap. So main still has to know some requirements, and on top of that they're no longer checked by the compiler (you get a runtime panic instead of a compile fail). It's not a great tradeoff.

Reply:
ok, I'll poke the example around and update the PR as per your suggestions so that at least there is an example for others even though it doesn't meet my initial requirements.